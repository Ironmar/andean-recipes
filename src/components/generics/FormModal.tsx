import React, { ChangeEvent, FormEvent, useContext, useState } from "react";
import AddIngredients from "./AddIngredients";
import { v4 as uuidv4 } from "uuid";

import addIcon from "../../assets/images/add_icon.svg";
import { IIngredient, IRecipe } from "../../interfaces/IRecipes";
import { RecipeContex } from "../../context/RecipeContex";
import { wordsRegx } from "../../constants/regExp";
import Error from "../generics/Error";

type Props = {
  setShowModal: (value: boolean) => void;
};

type inputElementType = React.ReactElement;
type eventChangeType = ChangeEvent<HTMLInputElement | HTMLTextAreaElement>;

const FormModal: React.FC<Props> = ({ setShowModal }) => {
  const recipeContex = useContext(RecipeContex);
  const [inputElement, setInputElement] = useState<inputElementType[]>([]);

  const [ingredientsList, setIngredientsList] = useState<IIngredient[]>([]);
  const [formValue, setFormValue] = useState<IRecipe>({
    name: "",
    ingredients: [],
    instructions: "",
  });
  const [errorWords, setErrorWords] = useState<boolean>(false);

  const { name, instructions } = formValue;

  const addIngredent = (value: IIngredient) => {
    setIngredientsList([...ingredientsList, value]);
  };

  const handleChange = ({ target }: eventChangeType) => {
    setFormValue({ ...formValue, [target.name]: target.value });
    const validator = wordsRegx.test(target.value);
    setErrorWords(validator);
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    formValue.id = uuidv4();
    recipeContex?.setRecipes([
      ...recipeContex.recipes,
      { ...formValue, ingredients: ingredientsList },
    ]);
    setShowModal(false);
  };

  return (
    <div className="modal">
      <form className="form" onSubmit={handleSubmit}>
        <div className="form-group-header">
          <label htmlFor="recipe">New Recipe</label>
          <input
            id="recipe"
            type="text"
            placeholder="Recipe Name"
            name="name"
            value={name}
            onChange={handleChange}
          />
          <button
            type="button"
            className="close-modal"
            onClick={() => setShowModal(false)}
          >
            X
          </button>
        </div>
        <div className="form-group">
          <h3>Ingredients</h3>
          {inputElement.map((element) => element)}
          <button
            className="modal-add-button"
            type="button"
            onClick={() =>
              setInputElement([
                ...inputElement,
                <AddIngredients key={uuidv4()} addIngredient={addIngredent} />,
              ])
            }
          >
            <img src={addIcon} alt="add_icon" />
          </button>
        </div>
        <div className="form-group">
          <h3>Instructions</h3>
          <textarea
            className="instructions"
            placeholder="Instructions"
            name="instructions"
            value={instructions}
            onChange={handleChange}
          ></textarea>
          {errorWords && <Error message="Incorrect word" />}
        </div>
        <button className="create-recipe " type="submit">
          âœ“ Create
        </button>
      </form>
    </div>
  );
};
export default FormModal;
