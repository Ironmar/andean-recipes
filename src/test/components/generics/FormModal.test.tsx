import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import toJson from "enzyme-to-json";

import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { createSerializer } from "enzyme-to-json";

import "@testing-library/jest-dom";
import FormModal from "../../../components/generics/FormModal";
configure({ adapter: new Adapter() });
//@ts-ignore
expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

describe("Test in <FormModal/>", () => {
  let wrapper: ShallowWrapper<any, Readonly<{}>>;
  const setShowModal = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<FormModal setShowModal={setShowModal} />);
  });
  test("should be return the component correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
