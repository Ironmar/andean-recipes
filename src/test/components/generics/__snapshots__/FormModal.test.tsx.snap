// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <FormModal/> should be return the component correctly 1`] = `
<div
  className="modal"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group-header"
    >
      <label
        htmlFor="recipe"
      >
        New Recipe
      </label>
      <input
        id="recipe"
        name="name"
        onChange={[Function]}
        placeholder="Recipe Name"
        type="text"
        value=""
      />
      <button
        className="close-modal"
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </div>
    <div
      className="form-group"
    >
      <h3>
        Ingredients
      </h3>
      <button
        className="modal-add-button"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="add_icon"
          src="add_icon.svg"
        />
      </button>
    </div>
    <div
      className="form-group"
    >
      <h3>
        Instructions
      </h3>
      <textarea
        className="instructions"
        name="instructions"
        onChange={[Function]}
        placeholder="Instructions"
        value=""
      />
    </div>
    <button
      className="create-recipe "
      type="submit"
    >
      âœ“ Create
    </button>
  </form>
</div>
`;
